rel/overlay/etc/default.ini

```diff
@@ -142,7 +142,7 @@ max_db_number_for_dbs_info_req = 100
 
 ;[jwt_auth]
 ; List of claims to validate
-; required_claims = exp
+; required_claims = exp,nbf
```

Comment:
I found this [answer on SO](https://stackoverflow.com/questions/43291659/usage-of-nbf-in-json-web-tokens) which talks about some possible use of `nbf`. How are we justifying this requirement?

Reply:
it's akin to the "Not Valid Before" and "Not Valid After" fields that all TLS certificates comes with, and recognises that tracking the "true" time is difficult, and that being able to influence clocks could lead to a security issue.

We check for expiration under the reasonable expectation that clocks typically move forward. If we don't check the nbf then there's an infinite window of time where the token is valid. If we check bnf then it's a specific timeframe, which is harder to target.

An alternative approach here that I considered, and would like your views on;

1) We validate every field present in the token, regardless of server setting. So, if the token includes `nbf`, `exp` or both, we'll validate them, rejecting them if they don't hold with respect to the servers clock.
2) The required_claims is then _only_ about whether we require those claims or not.
3) The default list would then be empty. Not insisting on exp or nbf but also not accepting a token with an nbf in the future or an exp in the past as valid.

Reply:
I would prefer that we require a minimum of `exp` via default.ini, so you'd have to explicitly remove it if you wanted an arguably less secure system.

Reply:
that was my original feeling for `exp,nbf` so I will add another commit tomorrow. Noting that before this PR the default of `exp` was ignored anyway.

Reply:
ah, you know, I think I will advocate for not including either -- because of replication. To use JWT with replication, you can set the "headers" directly and include your token, but this is no use if it expires. So, the token issuer can issue short-lived tokens for users (setting exp and/or nbf as they see fit) and we'll enforce them just because they're present. That same token issuer can issue a token without exp for use in replication tasks, though the only way to then revoke them is to change the secrets the server trusts...

Reply:
I think enforcing anything outside of the spec could surprise or get in the way of some legitimate design that somebody intends for.

My thought is to have the bare minimum out of a zero-configuration setup.  As in, if I run CouchDB and only enable JWTs, they should work.  This is more in the spirit of being easy to get started with and being able to dial in desired configuration incrementally.

That all said, I am open to some degree of caution, without forcing people to implement things that aren't applicable to their system (`nbf` being one potential example of such).

Reply:
we're discussing the defaults, @atrauzzi. Admins can change these settings. "Easy to get started" is my goal as long as it comports with "secure by default". What is your concrete suggestion for the required_claims default?

Reply:
Secure by default would be to require an `exp` claim in my opinion. :smile:

Reply:
An administrator that wishes to use JWT and replication would have to disable it, though.

Reply:
Ah, right!

Well, let's make sure the docs include something about changing the secret as a way to mitigate compromised tokens then.  Just so that people have some evidence that this scenario was considered.

Reply:
right. and of course you can use different secrets for the different key uses, so you could rotate your "replication" secret independently.

Reply:
> ah, you know, I think I will advocate for not including either -- because of replication. To use JWT with replication, you can set the "headers" directly and include your token, but this is no use if it expires. So, the token issuer can issue short-lived tokens for users (setting exp and/or nbf as they see fit) and we'll enforce them just because they're present. That same token issuer can issue a token without exp for use in replication tasks, though the only way to then revoke them is to change the secrets the server trusts...

I suspect someone using token based replication is also going to know how to edit default.ini, so would still vote +1 to include `exp`. FWIW, I think the replication token should still expire, even if it's set years in advance. An eternal token seems no better than a password.